File: data_structure/BIT/BIT.cpp
//Binary Indexed Tree
inline int lowbit(int n) {
    return n & -n;
}
template<typename T>
struct BIT {
    const static int maxn = MAXN;
    T s[maxn];	//s[0] is invalid
    int n;

    void Init(int _n) {
        n = _n;
        memset(s, 0, (n+1) * sizeof(s[0]));
    }
    void Init(int _n, const int a[]) {
        n = _n;
		s[0] = a[0];
		for (int i = 1; i <= n; ++i) {
			s[i] = s[i-1] + a[i];
		}
		for (int i = n; i; --i) {
			s[i] -= s[i - lowbit(i)];
        }
    }
	//x > 0
    void add(int x, T v) {
        for (; x <= n; x += lowbit(x))
            s[x] += v;
    }
    //The sum of [1, x]
    T sum(int x) {
        T ans = 0;
        for (; x; x ^= lowbit(x))
            ans += s[x];
        return ans;
    }
};

File: rust/bit.rs
fn lowbit(x: usize) -> usize {
    return x & (-(x as isize) as usize);
}
struct Bit {
    s: Vec<i32>
}
impl Bit {
    fn new(mut v: Vec<i32>) -> Bit {
        for i in 1..v.len() {
            v[i] += v[i-1];
        }
        for i in (0..v.len()).rev() {
            v[i] -= v[i - lowbit(i)];
        }
        Bit{ s: v }
    }
    fn add(&mut self, mut i: usize, v: i32) {
        while i < self.s.len() {
            self.s[i] += v;
            i += lowbit(i);
        }
    }
    // The sum of [1, i]
    fn sum(&self, mut i: usize) -> i32 {
        let mut ans = 0;
        while i != 0 {
            ans += self.s[i];
            i -= lowbit(i);
        }
        return ans;
    }
}

